File: compiler/bench/reverse.fs
\ expected: 499500
\ Reverse array benchmark - 10K iterations

1000 constant SIZE
10000 constant ITERS
create arr SIZE cells allot

: arr@ ( i -- n ) cells arr + @ ;
: arr! ( n i -- ) cells arr + ! ;

: swap-arr ( i j -- )
  over arr@ over arr@
  rot arr! swap arr! ;

: reverse-arr ( -- )
  0 SIZE 1-
  begin 2dup < while
    2dup swap-arr
    1 under+ 1-
  repeat
  2drop ;

: init-arr ( -- )
  SIZE 0 do i i arr! loop ;

: sum-arr ( -- n )
  0 SIZE 0 do i arr@ + loop ;

: bench ( -- sum )
  init-arr
  ITERS 0 do reverse-arr loop
  sum-arr ;

: main
  bench . cr ;

File: compiler/bench/reverse.c
// expected: 499500
// Reverse array benchmark - 10K iterations

#include <stdio.h>

#define SIZE 1000
#define ITERS 10000

int arr[SIZE];

void swap_arr(int i, int j) {
    int tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
}

void reverse_arr(void) {
    int i = 0, j = SIZE - 1;
    while (i < j) {
        swap_arr(i, j);
        i++;
        j--;
    }
}

void init_arr(void) {
    for (int i = 0; i < SIZE; i++) {
        arr[i] = i;
    }
}

long sum_arr(void) {
    long sum = 0;
    for (int i = 0; i < SIZE; i++) {
        sum += arr[i];
    }
    return sum;
}

long bench(void) {
    init_arr();
    for (int i = 0; i < ITERS; i++) {
        reverse_arr();
    }
    return sum_arr();
}

int main(void) {
    printf("%ld\n", bench());
    return 0;
}
